persistent/cPersistence.c

```diff
@@ -552,8 +554,20 @@ pickle___setstate__(PyObject *self, PyObject *state)
         }
 
         PyDict_Clear(*dict);
-        if (PyDict_Update(*dict, state) < 0)
-            return NULL;
+
+        i = 0;
+        while (PyDict_Next(state, &i, &d_key, &d_value)) {
+            /* normally the keys for instance attributes are
+               interned.  we should try to do that here. */
+            Py_INCREF(d_key);
```

Comment:
Still trying to wrap my head around the refcounting here...if I've got it right, I think Py_INCREF is here to convert the "borrowed" reference from PyDict_Next into an "owned" reference so that INTERN_INPLACE does not have a chance of triggering the object's finalizer. And then we no longer need to own the reference we pass to PyObject_SetItem because it will do its own incref for the reference stored in the object. Yes? If so I agree with your proposed change.